carta/util.py

```diff
@@ -129,3 +134,214 @@ def split_action_path(path):
+    }
+
+    SYMBOL_UNITS = {"", "'", "\""}
+    WORD_UNITS = NORMALIZED_UNIT.keys() - SYMBOL_UNITS
+
+    SYMBOL_UNIT_REGEX = rf"^(\d+(?:\.\d+)?)({'|'.join(SYMBOL_UNITS)})$"
+    WORD_UNIT_REGEX = rf"^(\d+(?:\.\d+)?)\s*({'|'.join(WORD_UNITS)})$"
+
+    @classmethod
+    def valid(cls, value):
+        """Whether the input string is a numeric value followed by an angular size unit.
+
+        A number without a unit is assumed to be in arcseconds. Permitted unit strings and their mappings to normalized units are stored in :obj:`carta.util.AngularSizeString.NORMALIZED_UNIT`. Whitespace is permitted after the number and before a unit which is a word, but not before ``'`` or ``"``.
+
+        Parameters
+        ----------
+        value : string
+            The input string.
+
+        Returns
+        -------
+        boolean
+            Whether the input string is an angular size.
+        """
+        return any((re.match(cls.WORD_UNIT_REGEX, value, re.IGNORECASE), re.match(cls.SYMBOL_UNIT_REGEX, value, re.IGNORECASE)))
+
+    @classmethod
+    def normalized(cls, value):
+        """Parse a string containing a numeric size and a unit, and return the size and the normalized unit.
+
+        A number without a unit is assumed to be in arcseconds. Permitted unit strings and their mappings to normalized units are stored in :obj:`carta.util.AngularSizeString.NORMALIZED_UNIT`. Whitespace is permitted after the number and before a unit which is a word, but not before ``'`` or ``"``.
+
+        Parameters
+        ----------
+        value : string
+            The string representation of the angular size.
+
+        Returns
+        -------
+        string
+            The numeric portion of the angular size string.
+        string
+            The normalized unit.
+
+        Raises
+        ------
+        ValueError
+            If the angular size string is not in a recognized format.
+        """
+        m = re.match(cls.WORD_UNIT_REGEX, value, re.IGNORECASE)
+        if m is None:
+            m = re.match(cls.SYMBOL_UNIT_REGEX, value, re.IGNORECASE)
+            if m is None:
+                raise ValueError(f"{repr(value)} is not in a recognized angular size format.")
+        value, unit = m.groups()
+        unit = cls.NORMALIZED_UNIT[unit]
+        return value, unit
+
+
+class WorldCoordinateString:
+    """Parses world coordinates."""
+
+    DEGREE_UNITS = {k for k, v in AngularSizeString.NORMALIZED_UNIT.items() if v == "deg"}
+
+    REGEX = {
+        "DEGREE_UNIT": rf"^-?(\d+(?:\.\d+)?)\s*({'|'.join(DEGREE_UNITS)})$",
+        "HMS_COLON": r"^-?\d{0,2}:\d{0,2}:(\d{1,2}(\.\d+)?)?$",
+        "HMS_LETTER": r"^(?:(-?\d{1,2})h)?(?:(\d{1,2})m)?(?:(\d{1,2}(?:\.\d+)?)s)?$",
+        "DMS_COLON": r"^-?\d*:\d{0,2}:(\d{1,2}(\.\d+)?)?$",
+        "DMS_LETTER": r"^(?:(-?\d+)d)?(?:(\d{1,2})m)?(?:(\d{1,2}(?:\.\d+)?)s)?$",
+        "DECIMAL": r"^-?\d+(\.\d+)?$",  # No units = degrees
+    }
+
+    @classmethod
+    def valid(cls, value):
```

Comment:
wonder if we need to validate value range? For example, for "hms" convention, `0 <= h < 24`, `0 <= m < 60`, and `0 <= s < 60`. For "dms", `-90 <= d <= 90`,`0 <= m < 60`, and `0 <= s < 60`. For "d" as "longitude", `0 <= d < 360` and as "latitude" `0 <= d <= 90`

Reply:
Also doable. We would need to be aware of both the coordinate system and which coordinate is which, so this probably needs to be done inside the function.

As far as I can tell, the frontend currently doesn't do any validation beyond the format -- it attempts to transform whatever values are provided and applies the result if it's valid.

Reply:
Actually, for HMS and DMS we can do a more specific check in the util class. We'll just need the system and the axes for the degree ranges.

Reply:
@kswang1029:

1. Should the latitude range for degrees not be `-90 <= d <= 90`?
2. For HMS and DMS, I assume  that e.g. H can only be 24 if M and S are zero (i.e. 24:12:34.5 = not valid). Is that correct?
3. What if the user sets DMS as the custom number format for a longitude axis? I understand that this is not the convention, but nothing is preventing it. Should we then validate that `0 <= d < 360`, as for decimal degrees?
4. Am I correct in assuming that longitude = right ascension = x axis and latitude = declination = y axis in all coordinate systems *except for ecliptic* in which case latitude = x axis and right ascension = y axis? And does it still make sense for the user to provide the coordinates in (x, y) order in the latter case?

Reply:
> @kswang1029:
> 
> 1. Should the latitude range for degrees not be `-90 <= d <= 90`?

When the coordinate system is galactic or ecliptic, we express the axis as Galactic/Ecliptic longitude and Galactic/Ecliptic latitude. Mostly, we use decimal degrees as the expression of longitude and latitude. For longitude, the range is 0 deg <= longitude < 360 deg, and for latitude, the range is -90 deg <= latitude <= 90 deg. 


> 2. For HMS and DMS, I assume  that e.g. H can only be 24 if M and S are zero (i.e. 24:12:34.5 = not valid). Is that correct?

Correct, the HMS (hour-minute-second) range is like 0h:0m:0s,  0h:0m:1s, ..., 23h59m59.999999....s. The DMS  (degree-arcmin-arcsec) range is -90d:0m:0s, ... -89d0m0s, ..., +90d0m0s. These are all sexagesimal. 

> 3. What if the user sets DMS as the custom number format for a longitude axis? I understand that this is not the convention, but nothing is preventing it. Should we then validate that `0 <= d < 360`, as for decimal degrees?

Usually for ICRS, FK5 and FK4 (ie equatorial system), we use the HMS-DMS convention (sexagesimal). But it is totally fine to convert sexagesimal format to decimal degree format. For GALACTIC and ECLIPTIC, we use decimal degrees convention. Rarely we use sexagesimal format for these two system (but it is valid to do such conversion).

> 4. Am I correct in assuming that longitude = right ascension = x axis and latitude = declination = y axis in all coordinate systems _except for ecliptic_ in which case latitude = x axis and right ascension = y axis? And does it still make sense for the user to provide the coordinates in (x, y) order in the latter case?

AST has its logic on where to map the longitude/latitude or RA/DEC axes to the canvas xy axes if the field has extra rotation. Assuming there is no extra field rotation, for ICRS, FK5 and FK4 systems, North is up and East is left. Then x is RA and y is DEC. For GALACTIC and ECLIPTIC systems without extra field rotation, North is up and East is left too. So x is longitude and y is latitude. The tricky part is, if we select GALACTIC grid to render from a native FK5 system, we will see the grid is rotated (hence we may see unexpected longitude/latitude to x/y mapping due to AST).

Reply:
Following up on our previous conversation, I also checked what happens in the GUI if you select HMS as the format for *latitude*. This appears to be valid, and negative values are converted to the expected negative HMS values, which suggests that in this case the range should be -6:00:00 to 6:00:00.

Reply:
Fortunately I was overthinking this. I see now that the orientation of X and Y *also* changes, so X is always longitude and Y is always latitude. That should make it easier to implement the ranges (and also to disable specific custom number options in the frontend).

Reply:
I also misread the HMS range -- I see that 24 is excluded from the hour range.